common/beerocks/bwl/common/key_value_parser.cpp

```diff
@@ -0,0 +1,225 @@
+        parsed_line_t parsed_line;
+        parse_line(ss_out, delimiter_list_out, parsed_line);
+        parsed_multiline.push_back(parsed_line);
+    }
+}
+
+void KeyValueParser::parse_event(std::string event_str, parsed_line_t &parsed_line)
+{
+    // Eliminate event log level from the begining of the event string : "<3>".
+    size_t idx_start = event_buffer_start_process_idx(event_str);
+
+    // An event string might start with keyless values such as Even Name.
+    // Find where the keyless parameters ends by finding where is the first parameter with key,
+    // assuming it has a preceding " <SomeKeyString>='".
+    auto idx = find_first_of_delimiter_pair(event_str, idx_start, ' ', '=');
+
+    // Put null terminator at the end of our key=val, for string stream construction.
+    if (idx != std::string::npos) {
+        event_str[idx] = '\0';
+    }
+
+    // Parse keyless params part and insert to the 'parsed_line' map object.
+    parse_event_keyless_params(event_str, idx_start, parsed_line);
+
+    // Fill the map with the rest of event data.
+    while (idx != std::string::npos) {
+
+        idx_start = ++idx;
+
+        // Find first '=' to skip on it
+        idx = event_str.find_first_of('=', idx_start);
+
+        // Find the next pair of delimiters index.
+        idx = find_first_of_delimiter_pair(event_str, ++idx, ' ', '=');
+
+        if (idx != std::string::npos) {
+            // Put null terminator at the end of our key=val, for ss_in construction.
+            event_str[idx] = '\0';
+        }
+
+        // Parse key=value
+        std::stringstream ss_in(event_str.c_str() + idx_start);
+        parse_line(ss_in, {'='}, parsed_line);
+    }
+}
+
+size_t KeyValueParser::event_buffer_start_process_idx(const std::string &event_str)
+{
+    // Eliminate event log level from the begining of the event string : "<3>"
+    auto idx_start = event_str.find_first_of(">");
+    if (idx_start != std::string::npos) {
+        idx_start++;
+    } else {
+        LOG(WARNING) << "Empty event! event_string: " << event_str;
+    }
+    return idx_start;
+}
+
+void KeyValueParser::parse_event_keyless_params(const std::string &event_str, size_t idx_start,
+                                                parsed_line_t &parsed_line)
+{
+    // Insert to map known prams without key
+    std::stringstream ss(event_str.c_str() + idx_start);
+    std::string str_storage;
+    bool opcode = true;
+    while (std::getline(ss, str_storage, ' ')) {
+        if (opcode) {
+            // assume that the second param is data or event name
+            parsed_line[EVENT_KEYLESS_PARAM_OPCODE] = str_storage;
+            opcode                                  = false;
+        } else if (beerocks::net::network_utils::is_valid_mac(str_storage)) {
+            parsed_line[EVENT_KEYLESS_PARAM_MAC] = str_storage;
+        } else if (!strncmp(str_storage.c_str(), "wlan", 4)) {
+            parsed_line[EVENT_KEYLESS_PARAM_IFACE] = str_storage;
+        }
```

Comment:
I've always hated this abomination as well...

It would make a whole lot more sense to me if parsed_line would have both a vector of keyless parameters, and a map of key-value parameters. And I don't think it makes any sense at all to duck-type the MAC address: either you expect a MAC address in that position and you parse it as such, or you expect something else and then you *should* parse it as a MAC address. Who knows, it might e.g. be an SSID that happens to be formatted as a MAC address...

Reply:
I'm sorry that you don't like this "abomination", but this solution is designed to be a single function that handles whatever output of hostapd.
Believe me, I would be happy if all of hostapd output was ordered on "key=value.." and I wouldn't have to do it. 

If you think you have a better implementation, feel free to change it. Just make sure it supports all of the hostapd output events.

Reply:
There are more keyless parameters other than MAC and interfaces, such as "freq'" in GET_FAILSAFE_CHAN event.
Look in ap/mon_wlan_hal_dwpal.cpp for all events with:
    FieldsToParse fieldsToParse[] = {
        {(void *)freq, &numOfValidArgs[0], DWPAL_STR_PARAM, **NULL**, sizeof(freq)},

NULL means keyless param, so it's not generic enough, need to think on another solution, you can look in dwpal parser to see how to identify them.

In addition, interface names can be other than "wlanX", in some platforms it can be "wlX" etc., so again, it's not generic enough.

Reply:
GET_FAILSAFE_CHAN is on the ap_dwapl and not the mon, but anyway it is not an "event" and therefore will not be parsed as an event on the code you gave the comments on.
In that case, there is nothing that can be done, except parsing it manually, and that what I did on the bwl_wav.
"multiap_common/browse/beerocks/bwl/wav/ap_wlan_hal_wav.cpp:
```cpp
bool ap_wlan_hal_wav::failsafe_channel_get(int &chan, int &bw)
{
    LOG(TRACE) << __func__;

    // Build command string
    std::string cmd = "GET_FAILSAFE_CHAN";
    char *reply;

    // Send command
    if (!wpa_ctrl_send_msg(cmd, &reply)) {
        LOG(ERROR) << "failsafe_channel_get() failed!";
        return false;
    }

    // Custom reply parsing
    if (!strncmp(reply, "UNSPECIFIED", 11)) {
        chan = -1;
        bw   = -1;
    } else if (!strncmp(reply, "ACS", 3)) {
        chan = bw = 0;

    } else {
        int freq;
        sscanf(reply, "%d %*s %*s bandwidth=%d", &freq, &bw);
        chan = beerocks::utils::wifi_freq_to_channel(freq);
    }

    return true;
}
```

This solution is not and can't be 100% proof. It was deisgned to handle the majority of cases, and the tiny part it could not handle has to be handled manually.
If it was 100% successful, it probably looked like the dwpal parser, which is something we definitely don't want.

Reply:
Let's say I'm a new developer, adding new event\cpmmand reply or anything to parse, how should I know if the parser fits my needs?
Probably I will look at a different event as an example, and see the parser being used, so I will use it as well right?

We can't rely on that developers should know how the parser works (as also know, not all of us know how dwpal parser works), and it's not a good approach sometimes to use the parser and sometimes not, it should be unified.

And if the dwpal parser can be generic enough, what's so wrong with it? 
As I see it now, if the C++ parser will not be unified for ALL parsing and will not be generic enough, it's not better than the dwpal parser!